qiskit_nature/transformers/active_space_transformer.py

```diff
@@ -0,0 +1,193 @@
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2021.
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+
+"""The Active-Space Reduction interface."""
+
+from typing import List, Optional
+import logging
+import numpy as np
+
+from qiskit_nature import QMolecule, QiskitNatureError
+from .base_transformer import BaseTransformer
+
+logger = logging.getLogger(__name__)
+
+
+class ActiveSpaceTransformer(BaseTransformer):
+    """The Active-Space reduction."""
+
+    def __init__(self, num_electrons: int, num_orbitals: int, num_alpha: Optional[int] = None,
+                 active_orbitals: Optional[List[int]] = None):
+        """Reduces the 1- and 2-electron integrals of a `QMolecule` into an active space.
+
+        Args:
+            num_electrons: the number of active electrons.
+            num_orbitals: the number of active orbitals.
+            num_alpha: the optional number of active alpha-spin electrons.
+            active_orbitals: A list of indices specifying the molecular orbitals of the active
+                             space. This argument must match with the remaining arguments and should
+                             only be used to enforce an active space that is not chosen purely
+                             around the Fermi level.
+        """
+        self.num_electrons = num_electrons
+        self.num_orbitals = num_orbitals
+        self.num_alpha = num_alpha
+        self.active_orbitals = active_orbitals
+
+    def transform(self, q_molecule: QMolecule) -> QMolecule:
```

Comment:
I was more considering  like these transforms to act on an h1/h2+num_particles and return say an extracted energy where h1/h2/num_particles would be updated according. Transforming the set of properties is then done by the class holding the set of them (as I described it here the class that is holding the raw driver result and has these integrals). This transformation needs to get applied to dipoles etc. And what is holding the extracted energies can maintain a list of them along with say some transform name so that the various extracted energies can be printed in some understandable way and correlated to the transforms

Reply:
As also mentioned above, I think this aspect will have to be dealt with in the later refactoring of the driver and `QMolecule` classes. Right now, I don't think it makes sense to add an auxiliary container with only the `h1` and `h2` integrals (which we are likely to refactor) if we can simply write all transformers to use the `QMolecule` directly. It does contain all of the information we need (just not in `h1` and `h2` fields) but readily exposes that format in the `one/two_body_integrals`.

Reply:
QMolecule currently does not contain the other operator/integral logic for spin, num particles etc, my thought was to have this in this additional class along with storing energies, h1/h2 etc. Yes it would be new and might have to be refactored later too - but its new stuff where ever it goes

Reply:
I kind of like what @dlasecki did over in #49 with regards to the additional operators. He created "calculators" which construct them based on the most minimal input. This is something I would have put into the "properties" or "observables" which we discussed at another point (but which we will leave open for the driver refactoring).
The `QMolecule` does contain all of the information which is necessary to construct all of the operators, it just does not include all of them in the `h1`/`h2` format.

Reply:
I did not believe h1/h2 would contain all the aspects - since num particles is altered and orbitals re-indexed etc. For me there is the set of things that are being acted upon, ie the above, and whether or not more info is needed, such as what atoms comprise the qmolecule in order that the core orbitals can be figured out for freeze core. Having access to data is one thing - maybe some of that data needed by transforms needs to be transformed too. I guess I was already thinking about keeping things apart since I was predicating the thought on all the transforms working on the block spin formatted h1/h2 which are not part of QMolecule. It seems the active space transform you have done on the integrals from which we computed the block spin form - ie. the integrals straight out the driver. I will note the transformation code we have today, that is in FermionicOperator or in the ParticleHole file works on blocked spin (well actually the particle hole logic internally was done on interleaved spin and the block->interleave and back to block was done internal to that code. Working on the original ints would be a little more efficient, its just not what has been done in the existing code.

Reply:
I think we can see how things develop on a case-by-case basis for the different transformers. If it is more efficient to operate on the raw driver integrals directly we can do that for now and see how to integrate everything nicely in the refactoring of the `QMolecule` class.